### ISSUE TITLE
Building Python bindings without collision support is not possible

### ISSUE THREAD

**Issue by Pouya-moh** [ADVRHumanoids/OpenSoT#216]
Hi!

## Context

As the title suggests, I'm trying to compile OpenSoT with collision support `off`, and python bindings `on`. This, however, seems to be not working. The reason we want to do this is that when `xbot2` is built with collision support it takes around 50 minutes and this interrupts our CICD's significantly. 

## Reproduce
Assuming `MatLogger2` and system dependencies are installed, from a clean environment:
```bash
source /opt/ros/humble/setup.bash

git clone https://github.com/Gauss-Robotics/xbot2_interface.git
cd xbot2_interface
mkdir build && cd build
cmake .. -DCMAKE_BUILD_TYPE=Release -DXBOT2_IFC_BUILD_COLLISION=OFF # note the cmake flag
make install -j 16
```

and then

``` bash
git clone https://github.com/Gauss-Robotics/OpenSoT.git
cd OpenSoT
mkdir build && cd build
cmake .. -DCMAKE_BUILD_TYPE=Release -DOPENSOT_COMPILE_COLLISION=OFF # this fails
make -j 16
```

What happens is that even though in 

https://github.com/ADVRHumanoids/OpenSoT/blob/50ce7e995ceddf7abdcee9b8c76d1a8ee2db8d71/CMakeLists.txt#L109-L113

the C++ is excluded, the python bindings are still considered

``` text
In file included from /root/src/OpenSoT/include/OpenSoT/constraints/velocity/CollisionAvoidance.h:24,
                 from /root/src/OpenSoT/bindings/python/constraints/velocity.hpp:7,
                 from /root/src/OpenSoT/bindings/python/pyopensot.cpp:10:
/usr/local/include/xbot2_interface/xbot2_interface/collision.h:5:2: error: #error "collision support not included: did you link your library against the xbot2_interface::collision cmake target?"
    5 | #error "collision support not included: did you link your library against the xbot2_interface::collision cmake target?"
      |  ^~~~~

```

## Potential workaround

We can do some macro magic or somehow break the python bindings in individual files. I personally don't like any of the solutions, that's why I trust you guys might have better ideas.

I hope this issue makes sense to you and many thanks!

**Comment by EnricoMingo** [comment]
Dear @Pouya-moh, I have (hopfully) fixed th bug in the branch `issue#216`. Could you please test it?
If working I will merge it in `4.0-devel`. Thank you!

**Comment by EnricoMingo** [comment]
Dear @Pouya-moh, any news on this?

**Comment by Pouya-moh** [comment]
Dear @EnricoMingo sorry for delayed reply. I tried it out in an isolated environment (ROS2 image) and it failed due to some of the lines you added:

```
Adding src/constraints/velocity/ConvexHull.cpp to compilation
Adding src/utils/convex_hull_utils.cpp to compilation
Internal qpOASES will be used!
Adding src/solvers/QPOasesBackEnd.cpp to compilation
-- Performing Test COMPILER_SUPPORTS_CXX11
-- Performing Test COMPILER_SUPPORTS_CXX11 - Success
-- Performing Test COMPILER_SUPPORTS_CXX0X
-- Performing Test COMPILER_SUPPORTS_CXX0X - Success
Adding src/solvers/eiQuadProgBackEnd.cpp to compilation
CMake Error at cmake/Modules/AddUninstallTarget.cmake:69 (add_custom_target):
  add_custom_target cannot create target "uninstall" because another target
  with the same name already exists.  The existing target is a custom target
  created in source directory "/OpenSoT".  See documentation for policy
  CMP0002 for more details.
Call Stack (most recent call first):
  CMakeLists.txt:319 (include)


**** Creating target release_deb to generate .deb for open_sot, to use it type: make release_deb ****
-- Will generate package 'open_sot-4.0.0-4dbd5bec-amd64-jammy.deb'
-- Performing Test HAS_FLTO
-- Performing Test HAS_FLTO - Success
-- check for pybind11
-- compiling python bindings
CMake Error at /usr/lib/cmake/pybind11/pybind11NewTools.cmake:207 (python3_add_library):
  Unknown CMake command "python3_add_library".
Call Stack (most recent call first):
  bindings/python/CMakeLists.txt:32 (pybind11_add_module)


-- Configuring incomplete, errors occurred!
```

Specifically the line `bindings/python/CMakeLists.txt:32` is part of the new changes. In parallel I will see if I can fix it myself. If I success I will let you know.

**Comment by Pouya-moh** [comment]
Updates: here are changes that I made and only after that I was able to configure cmake and build it successfully: 
- We had to comment the line below. I know it breaks your uninstall routine but I was not able to find where else the other target is set
   https://github.com/ADVRHumanoids/OpenSoT/blob/3e114c49eb04f222c72b702efec3d7dbce11c2e1/CMakeLists.txt#L316   
- I added `cmake_policy(SET CMP0057 NEW)` to the main cmake file but I think this is only due to some version incompability.
- I had to uncomment the line below, but also this is related to the version shipped with my docker image
  https://github.com/ADVRHumanoids/OpenSoT/blob/3e114c49eb04f222c72b702efec3d7dbce11c2e1/bindings/python/CMakeLists.txt#L7


After these changes the package compiles successfully.

**Comment by EnricoMingo** [comment]
This is very strange. Those lines were there also before, specifically in the ```4.0-devel``` branch.

**Comment by Pouya-moh** [comment]
@EnricoMingo You're right, but to build our fork which are based on ROS2 we had to do it. Could this be related to some stupid stuff from the `ament`? Do you have a containerized CI pipeline internally?

**Comment by EnricoMingo** [comment]
Nope, as well we do not have yet much tested opensot on ros2. If you comment 

`find_package(Python ${PYBIND11_PYTHON_VERSION} COMPONENTS Development Interpreter REQUIRED) `

you still have the bindings correctly compiled?

**Comment by EnricoMingo** [comment]
Can you confirm me that, despite the two parts commented, the changes of the branch permits you to compile without collision support?

**Comment by Pouya-moh** [comment]
Hi Enrico, sorry for long silences but I wanted to do a proper test :(

So I did a detail work from scratch and found some problems: if I comment `find_package(Python ${PYBIND11_PYTHON_VERSION} COMPONENTS Development Interpreter REQUIRED)` it compiles and links, **HOWEVER** if I open a python3 interpreter I get the following:

```python
>>> import pyopensot as pysot
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: /root/.local/lib/python3.10/site-packages/pyopensot.cpython-310-x86_64-linux-gnu.so: undefined symbol: _ZTIN7OpenSoT7solvers4HCODE

```

Interestingly, `OPENSOT_SOTH_FRONT_END` is set to `OFF` but somehow the bindings is still trying to link against it. Would it help if I prepare you a Dockerfile for rapid testing?

**Comment by EnricoMingo** [comment]
Hi @Pouya-moh, this is the same issue with the collision avoidance (and we could have it also for other flags...). I should have solved it in the latest commit, same branch. Could you test it in your docker, please? You will have still to comment `find_package(Python ${PYBIND11_PYTHON_VERSION} COMPONENTS Development Interpreter REQUIRED)` .

**Comment by Pouya-moh** [comment]
Hi @EnricoMingo. I can confirm your last changes solved the `OPENSOT_SOTH_FRONT_END` :partying_face: 
That being said, `find_package(Python ${PYBIND11_PYTHON_VERSION} COMPONENTS Development Interpreter REQUIRED)` still must remain uncommented. 

But I think this simply related to the fact with `humble` we are working with different versions of pybind(?). In any case it is working for us. Just a headsup, I have tested this with the changes mentioned above with ROS2 humble and 22.04. Thanks a lot!

**Comment by EnricoMingo** [comment]
Great, thank you, now the behavior of the CMakeLists of the bindings is more aligned with the main one. I have merged into `4.0-devel`. I will not delete the branch for the moment.
